File: lesson-8/java/Bike.java
/*
 * This is the bike class
 *
 * @author Mohamad Tanbari
 * @version 1.0.0
 * @since 2024-11-04
 */

/**
 * The Bike class.
 */
public class Bike extends Vehicle {
    /**
     * Number of wheels constant.
     */
    private static int numberOfWheels = 2;

    /**
     * The cadence property.
     */
    private int cadence;

    /**
     * The gear property.
     */
    private int gear = 1;

    /**
     * Constructor.
     *
     * @param color - The color you want the vehicle to start with.
     * @param maxSpeed - The max speed the vehicle can reach.
     */
    public Bike(
        String color,
        int maxSpeed) {
        // Set the properties
        super(color, maxSpeed, numberOfWheels);
        this.cadence = 0;
    }

    /**
     * The getter for the cadence property.
     *
     * @return The current cadence of the bike.
     */
    public int getCadence() {
        return this.cadence;
    }

    /**
     * The setter for the cadence property.
     *
     * @param newCadence - The number to set the cadence to.
     */
    public void setCadence(int newCadence) {
        if (newCadence < 0) {
            this.cadence = 0;
        } else {
            this.cadence = newCadence;
        }
    }

    /**
     * The getter for the gear property.
     *
     * @return The current gear of the bike.
     */
    public int getGear() {
        return this.gear;
    }

    /**
     * The setter for the gear property.
     *
     * @param newGear - The number to set the gear of the bike to.
     */
    public void setGear(int newGear) {
        if (newGear < 0) {
            this.gear = 0;
        } else {
            this.gear = newGear;
        }
    }

    /**
     * The status getter to show info about all properties of the vehicle.
     *
     * @return A string containing all the properties.
     */
    public String status() {
        final String status = "\n-> Speed: " + super.getSpeed()
                            + "\n-> Max Speed: " + super.getMaxSpeed()
                            + "\n-> Color: " + super.getColor()
                            + "\n-> Wheels: " + super.getNumberOfWheels()
                            + "\n-> Cadence: " + this.getCadence();

        return status;
    }

    /**
     * The accelerate method to increase the speed of the bike.
     *
     * @param appliedPower - The applied peddling power.
     * @param newGear - The new gear to set the bike to.
     */
    public void accelerate(int appliedPower, int newGear) {
        this.setCadence(this.getCadence() + appliedPower);
        super.setSpeed(super.getSpeed() + (this.getCadence() * this.getGear()));
    }

    /**
     * The ring bell method to ring the bike bell.
     *
     * @return The sound the bike makes.
     */
    public String ringBell() {
        final String sound = "Ding ding!\n";
        return sound;
    }
}

File: lesson-10/ts/Bike.ts
/*
 * The bike subclass extending Vehicle.
 *
 * Author: Mohamad Tanbari
 * Version: 1.0.0
 * Since: 2024-11-12
 */

import Vehicle from './Vehicle.ts'

const NUMBER_OF_WHEELS = 2

/**
 * The Bike class.
 */
export class Bike extends Vehicle {
  /**
   * The cadence property
   */
  private _cadence: number

  /**
   * The gear of the bike
   */
  private _gear: number = 1

  /**
   * Constructor.
   *
   * @param {string} color - The color of the bike.
   * @param {number} maxSpeed - The max speed of the bike.
   */
  constructor (color: string, maxSpeed: number) {
    super(color, maxSpeed, NUMBER_OF_WHEELS)
    this._cadence = 0
  }

  /**
   * The getter for the candence property.
   *
   * @return {number} The current cadence of the bike.
   */
  public get cadence (): number {
    return this._cadence
  }

  /**
   * The setter for the cadence property.
   *
   * @param {number} newCadence - The new cadence of the bike to set.
   */
  public set cadence (newCadence: number) {
    this._cadence = newCadence
  }

  /**
   * The getter for the gear property.
   *
   * @return {number} The current gear of the bike.
   */
  public get gear (): number {
    return this._gear
  }

  /**
   * The setter for the gear property.
   *
   * @param {number} newGear
   */
  public set gear (newGear: number) {
    if (newGear > 0) {
      this._gear = 1
    } else {
      this._gear = newGear
    }
  }

  /**
   * The accelerate method to override the one in Vehicle.
   *
   * @param {number} appliedPower
   * @param {number} newGear
   */
  public accelerate (appliedPower: number, newGear: number): void {
    // Update the gear
    this.gear = newGear
    this.cadence = this.cadence + appliedPower
    super.speed = super.speed + (this.cadence * newGear)
  }

  /**
   * The ring bell method to ring the bell of the bike.
   *
   * @return {string} the sound of the bell.
   */
  public ringBell (): string {
    return 'Ding ding!\n'
  }

  /**
   * This getter gives info on all properties of the vehicle.
   *
   * @return {string} A string containing all the property values.
   */
  public status (): string {
    const statusString = `-> Speed: ${super.speed}\n` +
                         `-> Max Speed: ${super.maxSpeed}\n` +
                         `-> Cadence: ${this.cadence}\n` +
                         `-> Wheels: ${super.numberOfWheels}\n` +
                         `-> Color: ${super.colour}`

    return statusString
  }
}
